File: examples/others/msg/lv_example_msg_3.c
#include "../../lv_examples.h"
#if LV_USE_MSG && LV_USE_SLIDER && LV_USE_LABEL && LV_BUILD_EXAMPLES

/*Define a message ID*/
#define MSG_INC             1
#define MSG_DEC             2
#define MSG_SET             3
#define MSG_UPDATE          4
#define MSG_UPDATE_REQUEST  5

static void value_handler(void * s, lv_msg_t * m);
static void value_handler(void * s, lv_msg_t * m);
static void btn_event_cb(lv_event_t * e);
static void label_event_cb(lv_event_t * e);
static void slider_event_cb(lv_event_t * e);

/**
 * Show how an increment button, a decrement button, as slider can set a value
 * and a label display it.
 * The current value (i.e. the system's state) is stored only in one static variable in a function
 * and no global variables are required.
 */
void lv_example_msg_3(void)
{

    lv_msg_subsribe(MSG_INC, value_handler, NULL);
    lv_msg_subsribe(MSG_DEC, value_handler, NULL);
    lv_msg_subsribe(MSG_SET, value_handler, NULL);
    lv_msg_subsribe(MSG_UPDATE, value_handler, NULL);
    lv_msg_subsribe(MSG_UPDATE_REQUEST, value_handler, NULL);

    lv_obj_t * panel = lv_obj_create(lv_scr_act());
    lv_obj_set_size(panel, 250, LV_SIZE_CONTENT);
    lv_obj_center(panel);
    lv_obj_set_flex_flow(panel, LV_FLEX_FLOW_ROW);
    lv_obj_set_flex_align(panel, LV_FLEX_ALIGN_SPACE_BETWEEN, LV_FLEX_ALIGN_CENTER, LV_FLEX_ALIGN_START);

    lv_obj_t * btn;
    lv_obj_t * label;

    /*Up button*/
    btn = lv_btn_create(panel);
    lv_obj_set_flex_grow(btn, 1);
    lv_obj_add_event_cb(btn, btn_event_cb, LV_EVENT_ALL, NULL);
    label = lv_label_create(btn);
    lv_label_set_text(label, LV_SYMBOL_LEFT);
    lv_obj_center(label);

    /*Current value*/
    label = lv_label_create(panel);
    lv_obj_set_flex_grow(label, 2);
    lv_obj_set_style_text_align(label, LV_TEXT_ALIGN_CENTER, 0);
    lv_label_set_text(label, "?");
    lv_msg_subsribe_obj(MSG_UPDATE, label, NULL);
    lv_obj_add_event_cb(label, label_event_cb, LV_EVENT_MSG_RECEIVED, NULL);

    /*Down button*/
    btn = lv_btn_create(panel);
    lv_obj_set_flex_grow(btn, 1);
    lv_obj_add_event_cb(btn, btn_event_cb, LV_EVENT_ALL, NULL);
    label = lv_label_create(btn);
    lv_label_set_text(label, LV_SYMBOL_RIGHT);
    lv_obj_center(label);

    /*Slider*/
    lv_obj_t * slider = lv_slider_create(panel);
    lv_obj_set_flex_grow(slider, 1);
    lv_obj_add_flag(slider, LV_OBJ_FLAG_FLEX_IN_NEW_TRACK);
    lv_obj_add_event_cb(slider, slider_event_cb, LV_EVENT_ALL, NULL);
    lv_msg_subsribe_obj(MSG_UPDATE, slider, NULL);

    /* As there are new UI elements that don't know the system's state
     * send an UPDATE REQUEST message which will trigger an UPDATE message with the current value*/
    lv_msg_send(MSG_UPDATE_REQUEST, NULL);
}

static void value_handler(void * s, lv_msg_t * m)
{
    LV_UNUSED(s);

    static int32_t value = 10;
    int32_t old_value = value;
    switch(lv_msg_get_id(m)) {
        case MSG_INC:
            if(value < 100) value++;
            break;
        case MSG_DEC:
            if(value > 0) value--;
            break;
        case MSG_SET: {
                const int32_t * new_value = lv_msg_get_payload(m);
                value = *new_value;
            }
            break;
        case MSG_UPDATE_REQUEST:
            lv_msg_send(MSG_UPDATE, &value);
            break;
        default:
            break;
    }

    if(value != old_value) {
        lv_msg_send(MSG_UPDATE, &value);
    }
}

static void btn_event_cb(lv_event_t * e)
{
    lv_obj_t * btn = lv_event_get_target(e);
    lv_event_code_t code = lv_event_get_code(e);
    if(code == LV_EVENT_CLICKED || code == LV_EVENT_LONG_PRESSED_REPEAT) {
        if(lv_obj_get_index(btn) == 0) {    /*First object is the dec. button*/
            lv_msg_send(MSG_DEC, NULL);
        }
        else {
            lv_msg_send(MSG_INC, NULL);
        }
    }
}

static void label_event_cb(lv_event_t * e)
{
    lv_obj_t * label = lv_event_get_target(e);
    lv_event_code_t code = lv_event_get_code(e);
    if(code == LV_EVENT_MSG_RECEIVED) {
        lv_msg_t * m = lv_event_get_msg(e);
        if(lv_msg_get_id(m) == MSG_UPDATE) {
            const int32_t * v = lv_msg_get_payload(m);
            lv_label_set_text_fmt(label, "%d %%", *v);
        }
    }
}

static void slider_event_cb(lv_event_t * e)
{
    lv_obj_t * slider = lv_event_get_target(e);
    lv_event_code_t code = lv_event_get_code(e);
    if(code == LV_EVENT_VALUE_CHANGED) {
        int32_t v = lv_slider_get_value(slider);
        lv_msg_send(MSG_SET, &v);
    }
    else if(code == LV_EVENT_MSG_RECEIVED) {
        lv_msg_t * m = lv_event_get_msg(e);
        if(lv_msg_get_id(m) == MSG_UPDATE) {
            const int32_t * v = lv_msg_get_payload(m);
            lv_slider_set_value(slider, *v, LV_ANIM_OFF);
        }
    }
}

#endif

File: examples/others/msg/lv_example_msg_3.py
# Define a message ID
MSG_INC            = 1
MSG_DEC            = 2
MSG_SET            = 3
MSG_UPDATE         = 4
MSG_UPDATE_REQUEST = 5

# Define the object that will be sent as msg payload
class NewValue:
    def __init__(self, value):
        self.value = value
    def __repr__(self):
        return f"{self.value} %"
class LV_Example_Msg_2:

    def __init__(self):
        self.value = 10
        lv.msg_subscribe(MSG_INC, self.value_handler, None)
        lv.msg_subscribe(MSG_DEC, self.value_handler, None)
        lv.msg_subscribe(MSG_SET, self.value_handler, None)
        lv.msg_subscribe(MSG_UPDATE, self.value_handler, None)
        lv.msg_subscribe(MSG_UPDATE_REQUEST, self.value_handler, None)

        panel = lv.obj(lv.scr_act())
        panel.set_size(250, lv.SIZE_CONTENT)
        panel.center()
        panel.set_flex_flow(lv.FLEX_FLOW.ROW)
        panel.set_flex_align(lv.FLEX_ALIGN.SPACE_BETWEEN, lv.FLEX_ALIGN.CENTER, lv.FLEX_ALIGN.START)

        # Up button
        btn = lv.btn(panel)
        btn.set_flex_grow(1)
        btn.add_event(self.btn_event_cb, lv.EVENT.ALL, None)
        label = lv.label(btn)
        label.set_text(lv.SYMBOL.LEFT)
        label.center()

        # Current value
        label = lv.label(panel)
        label.set_flex_grow(2)
        label.set_style_text_align(lv.TEXT_ALIGN.CENTER, 0)
        label.set_text("?")
        lv.msg_subscribe_obj(MSG_UPDATE, label, None)
        label.add_event(self.label_event_cb, lv.EVENT.MSG_RECEIVED, None)

        # Down button
        btn = lv.btn(panel)
        btn.set_flex_grow(1)
        btn.add_event(self.btn_event_cb, lv.EVENT.ALL, None)
        label = lv.label(btn)
        label.set_text(lv.SYMBOL.RIGHT)
        label.center()

        # Slider
        slider = lv.slider(panel)
        slider.set_flex_grow(1)
        slider.add_flag(lv.OBJ_FLAG_FLEX_IN_NEW_TRACK)
        slider.add_event(self.slider_event_cb, lv.EVENT.ALL, None)
        lv.msg_subscribe_obj(MSG_UPDATE, slider, None)


        # As there are new UI elements that don't know the system's state
        # send an UPDATE REQUEST message which will trigger an UPDATE message with the current value
        lv.msg_send(MSG_UPDATE_REQUEST, None)

    def value_handler(self,m):
        old_value = self.value
        id = m.get_id()
        if id == MSG_INC:
            if self.value < 100:
                self.value +=1
        elif id == MSG_DEC:
            if self.value > 0:
                self.value -=1
        elif id == MSG_SET:
            payload = m.get_payload()
            new_value=payload.__cast__()
            self.value = new_value.value
            # print("value_handler: new value: {:d}".format(new_value.value))
        elif id == MSG_UPDATE_REQUEST:
            lv.msg_send(MSG_UPDATE, NewValue(self.value))

        if self.value != old_value:
                lv.msg_send(MSG_UPDATE, NewValue(self.value));

    def btn_event_cb(self,e):
        btn = e.get_target_obj()
        code = e.get_code()
        if code == lv.EVENT.CLICKED or code == lv.EVENT.LONG_PRESSED_REPEAT:
            if btn.get_index() == 0:      # rst object is the dec. button
                lv.msg_send(MSG_DEC, None)
            else :
                lv.msg_send(MSG_INC, None)

    def label_event_cb(self,e):
        label = e.get_target_obj()
        code = e.get_code()
        if code == lv.EVENT.MSG_RECEIVED:
            m = e.get_msg()
            if m.get_id() == MSG_UPDATE:
                payload = m.get_payload()
                value=payload.__cast__()
                # print("label_event_cb: " + str(value))
                label.set_text(str(value))

    def slider_event_cb(self,e):
        slider = e.get_target_obj()
        code = e.get_code()
        if code == lv.EVENT.VALUE_CHANGED:
            v = slider.get_value()
            # print("slider_event_cb: {:d}".format(v))
            lv.msg_send(MSG_SET, NewValue(v))

        elif code == lv.EVENT.MSG_RECEIVED:
            m = e.get_msg()
            if m.get_id() == MSG_UPDATE:
                v = m.get_payload()
                value = v.__cast__()
                slider.set_value(value.value, lv.ANIM.OFF)

lv_example_msg_2 = LV_Example_Msg_2()

